# 8 Clusters

In [1]:
import json
from datasets import load_dataset

In [2]:
with open("/home/sohanp/sccl/results/SCCL.distilroberta.DEBERTA_V3_CLASS.explicit.wikitq_train.question.lr1e-05.lrscale100.contrastive.eta1.tmp0.5.alpha1.0.seed0/wikitq_cluster_assignment_1000.json", "r") as f:
    cluster_dict = json.load(f)

In [5]:
cluster_dict["0"]

{'question': 'what was the last year where this team was a part of the usl a-league?',
 'cluster_label': 7}

In [4]:
cluster_labels = {}

In [6]:
from tqdm import tqdm

In [7]:
for key, value in tqdm(cluster_dict.items(), position = 0, leave = True, total = len(list(cluster_dict.keys()))):
    if value["cluster_label"] not in list(cluster_labels.keys()):
        cluster_labels[value["cluster_label"]] = []

    cluster_labels[value["cluster_label"]].append(value["question"])

100%|██████████| 11321/11321 [00:00<00:00, 1573246.16it/s]


In [8]:
for key, val in cluster_labels.items():
    print(f"Questions in cluster {key} is {len(val)}")

Questions in cluster 7 is 969
Questions in cluster 4 is 1957
Questions in cluster 5 is 1381
Questions in cluster 1 is 1081
Questions in cluster 2 is 2157
Questions in cluster 6 is 1427
Questions in cluster 3 is 943
Questions in cluster 0 is 1406


In [24]:
cluster_labels[0][:20]

['how many consecutive friendly competitions did chalupny score in?',
 'how many finished all 225 laps?',
 'how many matches were in may 2010?',
 'how long was joseph black a coach?',
 'how many matches took place in the first premier?',
 'how many populations at most have any notes?',
 'how many top selling brands target dermatology?',
 "how many times was jim mcmanus jim osborne's partner?",
 'count how many of these members were unionists.',
 'how many academy awards have been won posthumously?',
 'the total number of helipads at hama military airport?',
 'how long was the marathon for camilla benjaminsson?',
 'how many games did the senators play in january?',
 'how many top scorers do not appear twice?',
 'how many teams had at least 15 wins?',
 'how many athletes are not ranked?',
 'how many of the contestants were students?',
 'how many spanish champions have there been?',
 'how many matches were held in the netherlands?',
 'how many people were born in 1976?']

In [18]:
cluster_labels[1][:20]

['how many people stayed at least 3 years in office?',
 'how many times did an italian cyclist win a round?',
 'how many total points did the bombers score against the bc lions?',
 'how many movies have links to other wikipedia pages about them?',
 'how many beta versions were released before the first full release?',
 'what number of games did new zealand win in 2010?',
 'how many games did at least 1500 people attend?',
 'how many songs charted above the 10th position on any chart?',
 'how many towns have a population higher than 1000?',
 'how many received a gold or silver medal in cycling?',
 'how many total points did russia win in this competition?',
 'how many defensive (df) players are there on the national team?',
 'number of goals manchester united scored against preston north end in the season',
 'how many times does "friendly" appear in the competition column?',
 'how many audio versions are less than five minutes long?',
 'how many articles were published in the 6th volume

In [19]:
cluster_labels[2][:20]

['who is the first away team on the chart',
 'which is the first city listed alphabetically?',
 'which model has the most in service?',
 'which name is first on the chart',
 'what film was released before "devakanya?"',
 'who has the most and best score?',
 'which club had the most losses?',
 'does theodis or david play center?',
 "what is the next model listed after disk'o?",
 'who directed the film rajanna?',
 'which driver appears the most?',
 'which ethnicity is previous from dungan',
 "what party has the most mp's?",
 'which rifle has the longest barrel?',
 'what is the total of conservatives',
 'which university has the least in endowment?',
 'which opponent has the most wins',
 'what property comes before tensile elongation?',
 'what two teams only have two titles?',
 'what week had the most attendance?']

In [20]:
cluster_labels[3][:20]

['in how many games did the winning team score more than 4 points?',
 'which album released by the band schnell fenster produced the most singles appearing on the australian peak chart?',
 'in 1996 the sc house of representatives had a republican majority. how many years had passed since the last time this happened?',
 'how many competitions had a score of 1-0 at most?',
 'how many miss northern ireland winners have made it to the top 30 placement at miss world since 2000?',
 'how many places list no zip code in either the lower or upper zip code?',
 'how many elections had at least a 0.2 percentage of constituency votes?',
 "at the women's 200 meter individual medley sm10 event at the 2012 summer paralympics, how long did it take aurelie rivard to finish?",
 'what style was the chateau de brissac rebuilt in after it was damaged during the french wars of religion?',
 'what year was the first to reach 1,000 or more live births?',
 'which election was the first to reach over 40% of the p

In [21]:
cluster_labels[4][:20]

["in what city did piotr's last 1st place finish occur?",
 'which ship in auckland had the fastest speed in knots?',
 'what counties had the least participants for the race?',
 'what club scored the only a total of 79 points?',
 'which kind of computer can most of the games be played on?',
 'in what year did miss pokhara last win the miss nepal award?',
 'which member of the 500 club has the least amount of home runs?',
 "which alumni in the 1990's has the least number of international caps?",
 'what title is at the top of the table?',
 'which month were the least amount of games held in?',
 'name someone else from scotland inducted before alan brazil.',
 'name the countries with the least amount if silver medals?',
 'other than parry, name an athlete from wales.',
 'name one county that only received 1,935 votes total.',
 'which district has the greatest total number of electorates?',
 'how many times, total, was the result "won"',
 'in which country did thierry tulasne win his last s

In [22]:
cluster_labels[5][:20]

['who was the opponent in the first game of the season?',
 'what was the first venue for the asian games?',
 'what was the only year keene won class aa?',
 "what's the number of parishes founded in the 1800s?",
 'what was the average number of years served by a coach?',
 'what is the last constellation featured on this table?',
 'who is the other person who is 24 years old besides reyna royo?',
 'who was the top winner in 2002 of the division 1 undergraduate?',
 'what is the total number of popular votes cast in 2003?',
 'what is the current total number of seats in the federal parliament?',
 'what is the total amount of senators i all departments?',
 'what is the most compilation albums released in one year?',
 'what was the number of days of the denver open?',
 'what is the least number of concerts given in a season',
 'other than ulm, what is the name of the other imperial city listed?',
 'what is the first airbase listed on the chart?',
 'what is the first party listed on this char

In [23]:
cluster_labels[6][:20]

['which is deeper, lake tuz or lake palas tuzla?',
 'which division three team also played in the division two season during the 1980s?',
 'who earned more medals--vietnam or indonesia?',
 'is the are of saint helena more than that of nightingale island?',
 'has the dominican republic won more or less medals than china?',
 'what vehicle maker other than dodge has the most vehicles in the roster?',
 'how long was the race in the all-africa games (distance)?',
 'who was the next ranked competitor after dior delophont and ligia grozav?',
 "the team's record in 2011 was the same was it's record in what year",
 'in 2008 in track and field events who broke more world records, usain bolt or haile gebrselassie?',
 'what country had the least gdp growth from 2007-2011?',
 'who is taller, the delegate from jalapa or from villa nueva?',
 'funningsfjørður and fuglafjørður are located on the same island as...?',
 'other than lake ercek, name a lake in van.',
 'what year did the album "jezebel" on b

In [25]:
cluster_labels[7][:20]

['what was the last year where this team was a part of the usl a-league?',
 'what is the difference in the number of temples between imabari and matsuyama?',
 'what is the total amount of processors that have both an integrated color display and an enhanced keyboard?',
 'what is the difference in league debut date between farrell and carne?',
 "what was whitney's best year for her album whitney in which she won the most rewards?",
 'what is the number of formula one series races that cochet has been in?',
 'what is the number of christian radio stations broadcasted by mxr yorkshire?',
 'what are the number of times antonov is listed as the manufacturer?',
 'who is the only person to score in the march 6 game against videoton this season?',
 'what are the number of times model 25 is listed on this chart?',
 'what was the best position achieved at the olympic games after the 1996 atlanta games?',
 'what is the total number of awards the lion king has won?',
 'what was her peak ranking in

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("/home/sohanp/sccl/datasets/wikitq_train.csv")

In [4]:
df["question"][0]

'what was the last year where this team was a part of the usl a-league?'

In [8]:
df["augmentation1"][0]

'what occurred the last year where this organisation was a part of the Dutch pro-league?'

In [7]:
df["augmentation2"][0]

'what was the last year when this squad was now part of the usl a - 2?'

# 5 Clusters

In [9]:
import json

In [10]:
from datasets import load_dataset

In [11]:
with open("/home/sohanp/sccl/results/SCCL.distilroberta.DEBERTA_V3_CLASS.explicit.wikitq_train.question.lr1e-05.lrscale100.contrastive.eta1.tmp0.5.alpha1.0.seed0.clusters5/wikitq_cluster_assignment_1000.json", "r") as f:
    cluster_dict = json.load(f)

In [13]:
train_dataset = load_dataset("wikitablequestions")["train"]

In [22]:
cluster_labels_separation_dict = {}

In [23]:
from tqdm import tqdm

In [41]:
cluster_labels = {}

In [42]:
for i, (key, value) in tqdm(enumerate(cluster_dict.items()), position = 0, leave = True, total = len(list(cluster_dict.keys()))):
    if value["cluster_label"] not in list(cluster_labels.keys()):
        cluster_labels[value["cluster_label"]] = {"id": [],
                                                  "question": [],
                                                  "table": [],
                                                  "answers": [],
                                                  "cluster_label": []}

    cluster_labels[value["cluster_label"]]["id"].append(i)
    cluster_labels[value["cluster_label"]]["question"].append(value["question"])
    cluster_labels[value["cluster_label"]]["table"].append(train_dataset[i]["table"])
    cluster_labels[value["cluster_label"]]["answers"].append(train_dataset[i]["answers"])
    cluster_labels[value["cluster_label"]]["cluster_label"].append(value["cluster_label"])

100%|██████████| 11321/11321 [00:04<00:00, 2275.31it/s]


In [50]:
import pandas as pd
for key, val in cluster_labels.items():
    df = pd.DataFrame(val)
    sampled_df = df.sample(frac = 0.025)
    sampled_df.reset_index(drop = True)

    sampled_df.to_csv(f"/home/sohanp/sccl/datasets/wikitq_train_cluster_{key}_sampled_0.25.csv", index = False)

### Examples from each cluster

In [35]:
cluster_labels[0][:20]

['how many people stayed at least 3 years in office?',
 'how many times did an italian cyclist win a round?',
 'how many total points did the bombers score against the bc lions?',
 'how many movies have links to other wikipedia pages about them?',
 'how many beta versions were released before the first full release?',
 'how many consecutive friendly competitions did chalupny score in?',
 'what number of games did new zealand win in 2010?',
 'how many games did at least 1500 people attend?',
 'how many matches were in may 2010?',
 'how many matches took place in the first premier?',
 'how many populations at most have any notes?',
 "how many times was jim mcmanus jim osborne's partner?",
 'how many songs charted above the 10th position on any chart?',
 'how many towns have a population higher than 1000?',
 'how many received a gold or silver medal in cycling?',
 'how many total points did russia win in this competition?',
 'how many defensive (df) players are there on the national team?

In [36]:
cluster_labels[1][:20]

['in how many games did the winning team score more than 4 points?',
 'which album released by the band schnell fenster produced the most singles appearing on the australian peak chart?',
 'in 1996 the sc house of representatives had a republican majority. how many years had passed since the last time this happened?',
 'how many competitions had a score of 1-0 at most?',
 'how many miss northern ireland winners have made it to the top 30 placement at miss world since 2000?',
 'how many places list no zip code in either the lower or upper zip code?',
 'how many elections had at least a 0.2 percentage of constituency votes?',
 "at the women's 200 meter individual medley sm10 event at the 2012 summer paralympics, how long did it take aurelie rivard to finish?",
 'what style was the chateau de brissac rebuilt in after it was damaged during the french wars of religion?',
 'what year was the first to reach 1,000 or more live births?',
 'in 2008 in track and field events who broke more world 

In [37]:
cluster_labels[2][:20]

["in what city did piotr's last 1st place finish occur?",
 'who was the opponent in the first game of the season?',
 'which is deeper, lake tuz or lake palas tuzla?',
 'what was the only year keene won class aa?',
 'which ship in auckland had the fastest speed in knots?',
 'what counties had the least participants for the race?',
 'what club scored the only a total of 79 points?',
 'which kind of computer can most of the games be played on?',
 'who is the other person who is 24 years old besides reyna royo?',
 'who was the top winner in 2002 of the division 1 undergraduate?',
 'in what year did miss pokhara last win the miss nepal award?',
 'which division three team also played in the division two season during the 1980s?',
 'who earned more medals--vietnam or indonesia?',
 'which member of the 500 club has the least amount of home runs?',
 'is the are of saint helena more than that of nightingale island?',
 "which alumni in the 1990's has the least number of international caps?",
 'w

In [38]:
cluster_labels[3][:20]

['who is the first away team on the chart',
 'what was the first venue for the asian games?',
 'which is the first city listed alphabetically?',
 'which model has the most in service?',
 "what's the number of parishes founded in the 1800s?",
 'which name is first on the chart',
 'what is the last constellation featured on this table?',
 'how many finished all 225 laps?',
 'what film was released before "devakanya?"',
 'who has the most and best score?',
 'which club had the most losses?',
 'does theodis or david play center?',
 "what is the next model listed after disk'o?",
 'what is the total amount of senators i all departments?',
 'who directed the film rajanna?',
 'what title is at the top of the table?',
 'how long was joseph black a coach?',
 'which driver appears the most?',
 'what is the most compilation albums released in one year?',
 'how many top selling brands target dermatology?']

In [39]:
cluster_labels[4][:20]

['what was the last year where this team was a part of the usl a-league?',
 'what is the difference in the number of temples between imabari and matsuyama?',
 'what was the average number of years served by a coach?',
 'what is the total amount of processors that have both an integrated color display and an enhanced keyboard?',
 'what is the total number of popular votes cast in 2003?',
 'what is the difference in league debut date between farrell and carne?',
 'what is the current total number of seats in the federal parliament?',
 "what was whitney's best year for her album whitney in which she won the most rewards?",
 'what is the number of formula one series races that cochet has been in?',
 'what is the number of christian radio stations broadcasted by mxr yorkshire?',
 'what are the number of times antonov is listed as the manufacturer?',
 'who is the only person to score in the march 6 game against videoton this season?',
 'what is the least number of concerts given in a season'

## Annotate reasons for each cluster

### Cluster 0


In [38]:
import pandas as pd

In [39]:
df = pd.read_csv("/home/sohanp/sccl/datasets/wikitq_train_cluster_0_sampled_0.25.csv")

In [79]:
question = df["question"][43]
table = eval(df["table"][43])
table_column_names = table["header"]
table_content_values = table["rows"]
table_df = pd.DataFrame.from_dict({str(col).lower(): [str(table_content_values[j][i]).lower() for j in range(len(table_content_values))] for i, col in enumerate(table_column_names)})
answer = df["answers"][43]

In [80]:
print(f"Question:\t {question}\n\n")
print(f"Answer:\t\t {answer}\n\n")
print(f"Table: ")
display(table_df)

Question:	 how many more bubas bison were released than copris hispanus linnaeus?


Answer:		 ['1319']


Table: 


species             country of origin  \
0                         bubas bison                 france, spain   
1                copris elphenor klug                  south africa   
2            copris hispanus linnaeus                         spain   
3      euoniticellus africanus harold                  south africa   
4          euoniticellus fulvus goeze                france, turkey   
5    euoniticellus intermedius reiche                  south africa   
6    euoniticellus pallipes fabricius                  iran, turkey   
7          geotrupes spiniger marsham                        france   
8       liatongus militaris castelanu     south africa (via hawaii)   
9             onitis alexis fabricius                  south africa   
10           onitis aygalus fabricius                  south africa   
11              onitis caffer boheman                  south africa   
12         onitis pecuarius lansberge                  south africa   
13     onitis vanderkelleni lansberge          kenya, rwanda, zaire   
14          onitis viridulus bohemann                  south africa   
15       onthophagus binodis thunberg                  south africa   
16      onthophagus gazella fabricius                  south africa   
17  onthophagus nigiventris d'orbigny                   east africa   
18               onthophagus obliquus       nigeria, senegal, zaire   
19            onthophagus sagittarius        sri lanka (via hawaii)   
20        onthophagus taurus schreber  spain, greece, italy, turkey   
21        sisyphus rubrus paschalidis                  south africa   
22                  sisyphus spinipes                  south africa   
23                                                           total:   

   total released   first release         last release  \
0           1,613      april 1983  1996 (not by csiro)   
1           2,287    january 1977             may 1983   
2             294    october 1983            june 1994   
3          49,009    october 1971        february 1984   
4          76,944      march 1978        february 1983   
5         248,637   november 1971        february 1984   
6          46,642      march 1977       september 1982   
7          12,082      april 1979        december 1983   
8          70,450    january 1968        november 1979   
9         186,441     august 1972        february 1984   
10         18,682    january 1977         january 1982   
11          8,738    october 1979           april 1984   
12         11,395   november 1976             may 1979   
13         10,852    october 1974        february 1982   
14          8,008  september 1976            july 1980   
15        173,018    october 1971        february 1982   
16        420,415   february 1968        february 1984   
17         29,960        may 1975           march 1983   
18          9,300    january 1976        november 1977   
19          9,075    january 1968           march 1977   
20        164,499   february 1975         january 1984   
21         85,933      march 1973        february 1980   
22         36,125      march 1972        december 1978   
23      1,680,399                                        

                                     areas of release  \
0                                                  wa   
1                                                 qld   
2                                                  wa   
3                          nsw, qld, sa, tas, vic, wa   
4                               nsw, sa, tas, vic, wa   
5                      act, nsw, nt, qld, sa, vic, wa   
6                                         nsw, sa, wa   
7                                  act, nsw, tas, vic   
8                                    nsw, nt, qld, wa   
9                                    nsw, nt, qld, wa   
10                                        nsw, sa, wa   
11                                                 wa   
12                                           nsw, qld   
13   

number of bubas bison released is 1,613. number of copris hispanus linnaeus released is 294. the difference is 1319.

### Cluster 1

In [81]:
import pandas as pd

In [82]:
df = pd.read_csv("/home/sohanp/sccl/datasets/wikitq_train_cluster_1_sampled_0.25.csv")

In [104]:
question = df["question"][35]
table = eval(df["table"][35])
table_column_names = table["header"]
table_content_values = table["rows"]
table_df = pd.DataFrame.from_dict({str(col).lower(): [str(table_content_values[j][i]).lower() for j in range(len(table_content_values))] for i, col in enumerate(table_column_names)})
answer = df["answers"][35]

In [105]:
print(f"Question:\t {question}\n\n")
print(f"Answer:\t\t {answer}\n\n")
print(f"Table: ")
display(table_df)

Question:	 in the 197172 national hurling league, who is the only single player ranked in the top 10 from laois?


Answer:		 ['Ger Cuddy']


Table: 


rank             player     county tally total opposition
0     1       éamonn flynn     dublin   3-5    14    wexford
1     1   charlie mccarthy       cork   2-8    14     galway
2     3         paddy fahy     galway   2-7    13    kildare
3     3       t. morrissey    wicklow   2-7    13      laois
4     5   charlie mccarthy       cork   3-3    12     offaly
5     5    michael moroney      clare   2-6    12     dublin
6     5      declan lovett      kerry   1-9    12     galway
7     5   charlie mccarthy       cork   1-9    12  tipperary
8     9          ger cuddy      laois   3-2    11    wicklow
9     9        pat enright  waterford   2-5    11      kerry
10    9        pat enright  waterford   1-8    11      laois
11    9       johnny walsh    kildare  0-11    11  waterford
12    9        eddie keher   kilkenny  0-11    11    wexford
13   14    brendan cummins       cork   3-1    10    wexford
14   14       p. j. molloy     galway   3-1    10      kerry
15   14         nick behan    kildare   3-1    10      kerry
16   14     seánie o'leary       cork   3-1    10   limerick
17   14    michael keating  tipperary   2-4    10    wexford
18   14    séamus richmond     antrim   2-4    10  westmeath
19   14      declan lovett      kerry   2-4    10  waterford
20   14    michael keating  tipperary   2-4    10     dublin
21   14        eddie keher   kilkenny   2-4    10     dublin
22   14          mick dwan    kildare   2-4    10     antrim
23   14   charlie mccarthy       cork   1-7    10   kilkenny
24   14         paddy fahy     galway  0-10    10      kerry
25   14       johnny walsh    kildare  0-10    10  westmeath
26   27            p flynn    wexford   3-0     9     offaly
27   27            n casey      clare   3-0     9       cork
28   27       willie moore   limerick   3-0     9       cork
29   27          ger cuddy      laois   3-0     9  westmeath
30   27           tom ryan     galway   3-0     9      laois
31   27     seánie o'leary       cork   3-0     9     offaly
32   27            p flynn    wexford   2-3     9       cork
33   27        eddie keher   kilkenny   2-3     9   limerick
34   27       johnny walsh    kildare   1-6     9     galway
35   27  francis loughnane  tipperary   1-6     9       cork
36   27      richie bennis   limerick   1-6     9       cork
37   27        pat enright  waterford   0-9     9    kildare
38   27       johnny walsh    kildare   0-9     9      kerry

from table, for country laois, there are 2 rows. however, for player ger cuddy from country laois, the rank is 9 which is in the top 10.

### Cluster 2

In [106]:
import pandas as pd

In [107]:
df = pd.read_csv("/home/sohanp/sccl/datasets/wikitq_train_cluster_2_sampled_0.25.csv")

In [211]:
question = df["question"][71]
table = eval(df["table"][71])
table_column_names = table["header"]
table_content_values = table["rows"]
table_df = pd.DataFrame.from_dict({str(col).lower(): [str(table_content_values[j][i]).lower() for j in range(len(table_content_values))] for i, col in enumerate(table_column_names)})
answer = df["answers"][71]

In [212]:
print(f"Question:\t {question}\n\n")
print(f"Answer:\t\t {answer}\n\n")
print(f"Table: ")
display(table_df)

Question:	 of all the retired drivers, which of them completed the least amount of laps?


Answer:		 ['Adrian Sutil']


Table: 


pos  no                driver          constructor laps time/retired grid  \
0     1   4         robert kubica           bmw sauber   70  1:36:24.227    2   
1     2   3         nick heidfeld           bmw sauber   70      +16.495    8   
2     3   9       david coulthard     red bull-renault   70      +23.352   13   
3     4  12            timo glock               toyota   70      +42.627   11   
4     5   2          felipe massa              ferrari   70      +43.934    6   
5     6  11          jarno trulli               toyota   70      +47.775   14   
6     7  17    rubens barrichello                honda   70      +53.597    9   
7     8  15      sebastian vettel   toro rosso-ferrari   70      +54.120   19   
8     9  23     heikki kovalainen     mclaren-mercedes   70      +54.433    7   
9    10   7          nico rosberg      williams-toyota   70      +54.749    5   
10   11  16         jenson button                honda   70    +1:07.540   20   
11   12  10           mark webber     red bull-renault   70    +1:11.299   10   
12   13  14    sébastien bourdais   toro rosso-ferrari   69       +1 lap   18   
13  ret  21  giancarlo fisichella  force india-ferrari   51     spun off   17   
14  ret   8       kazuki nakajima      williams-toyota   46     accident   12   
15  ret   5       fernando alonso              renault   44     spun off    4   
16  ret   6     nelson piquet jr.              renault   43       brakes   15   
17  ret   1        kimi räikkönen              ferrari   19    collision    3   
18  ret  22        lewis hamilton     mclaren-mercedes   19    collision    1   
19  ret  20          adrian sutil  force india-ferrari   13      gearbox   16   

   points  
0      10  
1       8  
2       6  
3       5  
4       4  
5       3  
6       2  
7       1  
8          
9          
10         
11         
12         
13         
14         
15         
16         
17         
18         
19

the minimum value of laps column is 13 and the corresponding driver is adrian sutil.

### Cluster 3

In [213]:
import pandas as pd

In [214]:
df = pd.read_csv("/home/sohanp/sccl/datasets/wikitq_train_cluster_3_sampled_0.25.csv")

In [360]:
question = df["question"][94]
table = eval(df["table"][94])
table_column_names = table["header"]
table_content_values = table["rows"]
table_df = pd.DataFrame.from_dict({str(col).lower(): [str(table_content_values[j][i]).lower() for j in range(len(table_content_values))] for i, col in enumerate(table_column_names)})
answer = df["answers"][94]

In [361]:
print(f"Question:\t {question}\n\n")
print(f"Answer:\t\t {answer}\n\n")
print(f"Table: ")
display(table_df)

Question:	 what material has a high thermal condition?


Answer:		 ['Moissanite']


Table: 


material                      formula  \
0              diamond                            c   
1              glasses  silica with pb, al, &/or tl   
2       white sapphire                        al2o3   
3               spinel                    mgo·al2o3   
4               rutile                         tio2   
5   strontium titanate                       srtio3   
6                  yag                     y3al5o12   
7                  ggg                    gd3ga5o12   
8       cubic zirconia          zro2(+ rare earths)   
9           moissanite                          sic   
10              quartz                       silica   

   refractive\nindex(es)\n589.3 nm dispersion\n431 – 687 nm  \
0                            2.417                    0.044   
1                            ~ 1.6                  > 0.020   
2                    1.762 – 1.770                    0.018   
3                            1.727                    0.020   
4                       2.62 – 2.9                     0.33   
5                             2.41                     0.19   
6                             1.83                    0.028   
7                             1.97                    0.045   
8                            ~ 2.2                   ~ 0.06   
9                    2.648 – 2.691                    0.104   
10                   1.543 – 1.554                            

   hardness\n(mohs'\nscale) density\n(g/cm3) thermal\ncond. state of\nthe art  
0                        10             3.52      excellent         (natural)  
1                       < 6        2.4 – 4.2           poor            1700 –  
2                         9             3.97           poor         1900–1947  
3                         8            ~ 3.6           poor         1920–1947  
4                       ~ 6             4.25           poor         1947–1955  
5                       5.5             5.13           poor         1955–1970  
6                      8.25      4.55 – 4.65           poor         1970–1975  
7                         7             7.02           poor         1973–1975  
8                     ~ 8.3            ~ 5.7           poor            1976 –  
9                  8.5–9.25              3.2           high            1998 –  
10                       7-      2.50 – 2.65                          ancient

from table, considering column themal cond., we see that the word high appears in one row and the corresponding material in that row is moissanite.

### Cluster 4

In [362]:
import pandas as pd

In [363]:
df = pd.read_csv("/home/sohanp/sccl/datasets/wikitq_train_cluster_4_sampled_0.25.csv")

In [387]:
question = df["question"][35]
table = eval(df["table"][35])
table_column_names = table["header"]
table_content_values = table["rows"]
table_df = pd.DataFrame.from_dict({str(col).lower(): [str(table_content_values[j][i]).lower() for j in range(len(table_content_values))] for i, col in enumerate(table_column_names)})
answer = df["answers"][35]

In [388]:
print(f"Question:\t {question}\n\n")
print(f"Answer:\t\t {answer}\n\n")
print(f"Table: ")
display(table_df)

Question:	 what is the number of home games the new york rangers won in january 1931?


Answer:		 ['1']


Table: 


game december               opponent     score record
0    9        4   @ montreal canadiens     5 - 4  3-4-2
1   10        6  @ toronto maple leafs     4 - 2  3-5-2
2   11        9        ottawa senators     3 - 2  4-5-2
3   12       14        detroit falcons     3 - 0  5-5-2
4   13       18          boston bruins     4 - 2  5-6-2
5   14       20        @ boston bruins  2 - 2 ot  5-6-3
6   15       23     montreal canadiens     5 - 1  6-6-3
7   16       25      @ ottawa senators     4 - 1  7-6-3
8   17       28   philadelphia quakers     4 - 2  8-6-3
9   18       30   @ new york americans  2 - 2 ot  8-6-4

looking at opponent column, we infer that presence of @ indicates an away game. so, first we need to retrieve rows where @ is not present in the opponent column. from table, there are 5 such rows. out of these rows, we need to compare the the score and report number of games won by new york rangers.